datafusion-cli/src/catalog.rs

```diff
@@ -151,10 +154,35 @@ impl SchemaProvider for DynamicFileSchemaProvider {
         // if the inner schema provider didn't have a table by
         // that name, try to treat it as a listing table
         let state = self.state.upgrade()?.read().clone();
-        let config = ListingTableConfig::new(ListingTableUrl::parse(name).ok()?)
+        let table_url = ListingTableUrl::parse(name).ok()?;
+
+        // Assure the `http` store for this url is registered if this
+        // is an `http(s)` listing
+        // TODO: support for other types, e.g. `s3`, may need to be added
+        match table_url.scheme() {
+            "http" | "https" => {
+                let url: &Url = table_url.as_ref();
+                match state.runtime_env().object_store_registry.get_store(url) {
+                    Ok(_) => {}
+                    Err(_) => {
```

Comment:
please, why the http client gets built on get_store error ?

Reply:
I saw it an easy way to avoid re-registering the store for the same url. If you query the same url multiple times, it will err on first call so the store will be registered, on subsequent queries it will not. Happy to change if you can suggest something different

Reply:
I agree this is confusing but follows the existing pattern in datafusion-cli, which is basically to defer registering the object store instance until something actually tries to access it (so, for example, we don't get S3 config errors b/c the config isn't setup if the user isn't actually querying S3)

As written this code will support http from select, which is great, but it doesn't support other url types (like `s3://...` etc. 

Instead of creating an HttpBuilder directly, would it be possible to call

https://github.com/apache/arrow-datafusion/blob/10ae9343368a893012aa80b66c02d45b4f461f9f/datafusion-cli/src/exec.rs#L305-L339

So all the types of object stores are covered? 

It would be great to add additional comments explaining the deferred registration logic, but i would be happy to do that as a follow on PR as well

Reply:
I'd like to follow up with a PR for other protocols so will try to re-use the existing logic and add more comments